File: surf-cloud-core/src/main/kotlin/dev/slne/surf/cloud/core/netty/client/source/NettyClientSourceImpl.java
package dev.slne.surf.cloud.core.netty.client.source;

import dev.slne.surf.cloud.api.netty.source.NettyClientSource;
import dev.slne.surf.cloud.core.netty.AbstractNettyBase;
import dev.slne.surf.cloud.core.netty.common.source.AbstractProxiedNettySource;
import io.netty.channel.Channel;
import java.util.Objects;
import lombok.Getter;
import lombok.experimental.Accessors;

@Getter
@Accessors(fluent = true)
public class NettyClientSourceImpl extends AbstractProxiedNettySource<NettyClientSource> implements NettyClientSource {

  private final Channel channel;

  public NettyClientSourceImpl(AbstractNettyBase<?, ?, NettyClientSource> base, Channel channel) {
    super(base);
    this.channel = channel;
  }

  @Override
  public final boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof NettyClientSourceImpl that)) {
      return false;
    }
    if (!super.equals(o)) {
      return false;
    }

    return Objects.equals(channel, that.channel);
  }

  @Override
  public int hashCode() {
    int result = super.hashCode();
    result = 31 * result + Objects.hashCode(channel);
    return result;
  }
}

File: surf-cloud-core/src/main/kotlin/dev/slne/surf/cloud/core/netty/client/source/NettyClientSourceImpl.kt
package dev.slne.surf.cloud.core.netty.client.source

import dev.slne.surf.cloud.api.netty.source.NettyClientSource
import dev.slne.surf.cloud.core.netty.AbstractNettyBase
import dev.slne.surf.cloud.core.netty.common.source.AbstractProxiedNettySource
import dev.slne.surf.cloud.core.netty.protocol.packet.handler.ServerRunningPacketListener
import io.netty.channel.Channel


class NettyClientSourceImpl(
    base: AbstractNettyBase<*, *, NettyClientSource>,
    override val channel: Channel
) : AbstractProxiedNettySource<NettyClientSource>(base), NettyClientSource {
    private var _connection: ServerRunningPacketListener? = null
    val connection get() = _connection ?: error("Connection not set")

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is NettyClientSourceImpl) return false
        if (!super.equals(other)) return false

        if (channel != other.channel) return false

        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + channel.hashCode()
        return result
    }

    override fun toString(): String {
        return "NettyClientSourceImpl(channel=$channel) ${super.toString()}"
    }
}
